core/domain/exp_jobs_one_off.py

```diff
@@ -216,68 +256,133 @@ def map(item):
-                                validity = 'Valid Expression'
+                                validity = 'Valid Algebraic Expression'
+                            elif is_valid_math_expression(
+                                    rule_input, algebraic=False):
+                                validity = 'Valid Numeric Expression'
                             elif is_valid_math_equation(rule_input):
-                                validity = 'Valid Equation'
+                                validity = 'Valid Math Equation'
+
+                            if validity != 'Invalid':
+                                types_of_inputs.add(validity)
+
+                            if len(types_of_inputs) > 1:
+                                yield ('ERROR', (
+                                    'The exploration with ID: %s and state '
+                                    'name: %s contains inputs that correspond '
+                                    'to two different types. Please resolve '
+                                    'this before running the upgradation '
+                                    'job.' % (item.id, state_name)))
+
+                            output_values = '%s %s: %s' % (
+                                item.id, state_name, rule_input)
+                            yield (validity, output_values.encode('utf-8'))
+
+    @staticmethod
+    def reduce(key, values):
+        if key.startswith('Valid'):
+            yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
+        else:
+            if key == 'Invalid' and len(values) != 0:
+                yield ('ERROR', (
+                    'There are some invalid inputs that need to be resolved '
+                    'before running the upgradation job. The invalid inputs '
+                    'are as follows:'))
+            yield (key, values)
+
+
+class MathExpressionUpgradationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration.
+
+    This validation is done by the 'is_valid_math_expression' or the
+    'is_valid_math_equation' function present in schema_utils.py.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        is_valid_math_expression = schema_utils.get_validator(
+            'is_valid_math_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+
+        if not item.deleted:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+            for state_name, state in exploration.states.items():
+                if state.interaction.id == 'MathExpressionInput':
+                    new_answer_groups = []
+                    new_interaction_id = ''
```

Comment:
You should set this to a valid value or at least validate at the end that it is not ''.

Reply:
Done.

Reply:
> Not done. I think there might be some confusion about what I meant by "validate". What I'm trying to say is that you should fail visibly somehow if this is still '' at the end, because presumably that is a non-expected case. Currently that case seems to be just swept under the rug.

Understood, I've added an explicit ERROR message for that case.

>  Also why is this ''? Shouldn't it be None?

I've changed the code so that it is consistent in both the jobs and easier to follow, which is why this is not required.